### ISSUE TITLE
Adding generics support 

### ISSUE THREAD

**Issue by Kocal** [KnpLabs/DoctrineBehaviors#516]
Hi, :hand:  

This is a feature request for adding generics types for PHPStan (or Psalm) on `Translatable`/`Translation` interfaces:
- https://medium.com/@ondrejmirtes/generics-in-php-using-phpdocs-14e7301953
- https://arnaud.le-blanc.net/post/phpstan-generics.html

Given the following classes:
```php
<?php

namespace App\Entity;

use \Knp\DoctrineBehaviors\Contract\Entity\TranslatableInterface;
use \Knp\DoctrineBehaviors\Contract\Entity\TranslationInterface;
use \Knp\DoctrineBehaviors\Model\Translatable\TranslatableTrait;
use \Knp\DoctrineBehaviors\Model\Translatable\TranslationTrait;

/**
 * @implements TranslatableInterface<ArticleTranslation>
 */
class Article implements TranslatableInterface {
    use TranslatableTrait;
}

/**
 * @implements TranslationInterface<Article>
 */
class ArticleTranslation implements TranslationInterface {
    use TranslationTrait;

    public function getTitle() { /* ... */ }
    public function setTitle(string $title) { /* ... */ }
}

class FooBar {}
```

By using generic types, we will improve static analyze and developer experience (when IDEs will support PHPStan/Psalm generics syntax):

```php
<?php

$article = new Article();

$article->getTranslations(); // typed to a Collection of `ArticleTranslation`
$article->getNewTranslations(); // typed to a Collection of `ArticleTranslation`

$article->translate('fr'); // typed to `ArticleTranslation`
$article->translate('fr')->setTitle('Mon titre'); // ok
$article->translate('fr')->setFooBar('Mon titre'); // fail

$article->addTranslation(new ArticleTranslation); // ok
$article->addTranslation(new FooBar); // fail

$article->removeTranslation(new ArticleTranslation); // ok
$article->removeTranslation(new FooBar); // fail

$article->setTranslations(new ArrayCollection([new ArticleTranslation]); // ok
$article->setTranslations(new ArrayCollection([new ArticleTranslation, new FooBar]); // fail

$articleTranslation = /* ... */;
$articleTranslation->getTranslatable(); // typed to `Article`
```

---

On one of our project at work, I've started to implement generics myself (on `translate` method, with new traits and interfaces which extend code from vendor) but it didn't work properly. 

I suspect PHPStan to use PHP's return type from [here](https://github.com/KnpLabs/DoctrineBehaviors/blob/master/src/Contract/Entity/TranslatableInterface.php#L33) since there is no PHPDoc annotations on the method. So a solution is to update this library for generics support.

As a current workaround, I've created 3 PHPStan extensions for the following methods, I can copy/paste them here if someone is interested:
- `TranslatableInterface#getTranslations`, `TranslatableInterface#getNewTranslations`
- `TranslatableInterface#translate`
- `TranslationInterface#getTranslatable`

---

Using generic types on `TranslatableInterface` would look like this:

```php
<?php

declare(strict_types=1);

namespace Knp\DoctrineBehaviors\Contract\Entity;

use Doctrine\Common\Collections\Collection;

/**
 * @phpstan-template T of TranslationInterface
 */
interface TranslatableInterface
{
    /**
     * @phpstan-return Collection<int,T>|T[]
     * @return Collection|TranslationInterface[]
     */
    public function getTranslations();

    /**
     * @phpstan-return Collection<int,T>|T[]
     * @return Collection|TranslationInterface[]
     */
    public function getNewTranslations();

    /**
     * @phpstan-param T $translation
     */
    public function addTranslation(TranslationInterface $translation): void;

    /**
     * @phpstan-param T $translation
     */
    public function removeTranslation(TranslationInterface $translation): void;

    /**
     * Returns translation for specific locale (creates new one if doesn't exists).
     * If requested translation doesn't exist, it will first try to fallback default locale
     * If any translation doesn't exist, it will be added to newTranslations collection.
     * In order to persist new translations, call mergeNewTranslations method, before flush
     *
     * @param string $locale The locale (en, ru, fr) | null If null, will try with current locale
     * @phpstan-return T
     */
    public function translate(?string $locale = null, bool $fallbackToDefault = true): TranslationInterface;

     /* ... */
}
```

I didn't test it but it may works. 

What do you think?

Thanks!

**Comment by TomasVotruba** [comment]
Sure :+1: , send PR

**Comment by TomasVotruba** [comment]
Closed by #520
